### ISSUE TITLE
pulse_meter seems to have broken with 2023.8.0

### ISSUE THREAD

**Issue by egite** [esphome/issues#4807]
### The problem

Greetings.

Upgrading from 2023.7.1 to 8.0, the pulse_meter function seems to have broken/changed.  The below now causes the counter to go crazy and increment at a high rate endlessly.  Removing the internal_filter causes the initial count after reset to be a randomly high number.  The pulse count afterwards counts, though is several orders of magnitude higher than actual.  Removing the debounce filters also causes endless increments.

I have a separate ESP8266 on an identical but separate water meter running the same code that no longer reports its count.  The sensor shows as "unknown".

Please let me know how I can help.

### Which version of ESPHome has the issue?

 2023.8.0

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

 2023.8.2

### What platform are you using?

ESP8266

### Example YAML snippet

```yaml
sensor:    
  - platform: pulse_meter
    pin: 
      number: GPIO5  #D1
      mode:  
        input: true
    name: "Pulse Meter"
    unit_of_measurement: "gal/min"
    internal_filter: 50ms   
    internal_filter_mode: EDGE
    accuracy_decimals: 1
    timeout: 10sec
    icon: "mdi:gauge"    
    filters: 
     - multiply: 0.0748052    
     - debounce: 50ms      
    total:
      name: "Total"
      accuracy_decimals: 3
      unit_of_measurement: "gal"
      device_class: water
      state_class: total_increasing
      filters: 
        - multiply: 0.0748052
        - debounce: 50ms
```

**Comment by Megu** [comment]
Got fairly similar issue. "Unknown" value on ESP32 using 2023.8.0 and 2023.8.1 versions of ESPHome addon on Home Assistant 2023.8.2. No issues running 2023.7.1 version of ESPHome.
```yaml
sensor:
    - platform: pulse_meter
      pin: GPIO19
      unit_of_measurement: 'kW'
      id: power1
      name: 'Power 1'
      device_class: "power"
      filters:
        - multiply: 0.06  # (60s/1000 pulses per kWh)
      internal_filter: $pulse_debounce
      total:
        id: electricity1
        name: Electricity 1
        unit_of_measurement: "kWh"
        accuracy_decimals: 3
        device_class: "energy"
        state_class: "total_increasing"
        filters:
          - multiply: 0.001```

**Comment by slallemand** [comment]
Same issue as @Megu ; with version 2023.8 I have a "Unknown" value on my ESP8266.
Reverted back to esphome 2023.7 and everything went back to normal and the pulse_meter is ok;

**Comment by pmartyniuk** [comment]
In addition, it is impossible to reset the counter using: **pulse_meter.set_total_pulses**
I just tested this and resetting the counter again does not work

**Comment by fabianschaefer** [comment]
I also have another issue with pulse meter since the [refactor](https://github.com/esphome/esphome/pull/4231). The "no pulse detected code now always ignores the last pulse. In my case I have the timeout set to 10s. I get a pulse at 01:44:05 and another at 01:44:09, but at 01:44:15 (10 seconds after the penultimate pulse), I get the "No pulse detected for 10s message" and the sensor goes to 0, which is wrong:
```[01:44:05][D][sensor:094]: 'water_meter_flow': Sending state 18.74292 L/min with 0 decimals of accuracy
[01:44:09][D][pulse_meter:057]: No pulse detected for 10s, assuming 0 pulses/min
[01:44:09][D][sensor:094]: 'water_meter_flow': Sending state 0.00000 L/min with 0 decimals of accuracy
[01:44:11][D][sensor:094]: 'water_meter_total': Sending state 38.00000 L with 0 decimals of accuracy
[01:44:15][D][pulse_meter:057]: No pulse detected for 10s, assuming 0 pulses/min
[01:44:15][D][sensor:094]: 'water_meter_flow': Sending state 0.00000 L/min with 0 decimals of accuracy
[01:44:17][D][sensor:094]: 'water_meter_total': Sending state 39.00000 L with 0 decimals of accuracy
[01:44:21][D][pulse_meter:057]: No pulse detected for 10s, assuming 0 pulses/min
[01:44:21][D][sensor:094]: 'water_meter_flow': Sending state 0.00000 L/min with 0 decimals of accuracy
[01:44:23][D][sensor:094]: 'water_meter_total': Sending state 40.00000 L with 0 decimals of accuracy
[01:44:27][D][pulse_meter:057]: No pulse detected for 10s, assuming 0 pulses/min```

**Comment by jmoery** [comment]
I'm also having those issues on 2023.8.5.  I'm using a pulse meter with a Shelly rgbw2 A/C input sensor to track switch presses.  I get an unknown clause as an initial state and then the pulse meter doesn't reset on the timeout interval when pulses aren't detected.

**Comment by fabianschaefer** [comment]
@jesserockz can you have a look here, part of the pulse counter integration is broken since the [refactor](https://github.com/esphome/esphome/pull/4231) has been merged. Especially the "no pulse detected" mechanism doesn't behave as expected anymore (see my last comment).

**Comment by CestLaGalere** [comment]
problem having installed 2023.8.2 - power usage going from about 0.4kW to 170kW!
The total has roughly tripled from about 20kWh / day to about 60kWh/day.
No changes to yaml, system updates (core to 2023.8.3 & ESPHome to 2023.8.2)

```
  - platform: pulse_meter
    pin: GPIO13
    unit_of_measurement: 'kW'
    accuracy_decimals: 3
    name: 'Pulse Power'
    id: pulse_mtr
    filters:
      - multiply: 0.06  # (60s/1000 pulses per kWh)

    total:
      unit_of_measurement: 'kWh'
      name: 'pulse Total Energy'
      accuracy_decimals: 2
      filters:
        - multiply: 0.001  # (1/1000 pulses per kWh)
```

**Comment by Grafton1046** [comment]
> I also have another issue with pulse meter since the [refactor](https://github.com/esphome/esphome/pull/4231). The "no pulse detected" code now always ignores the last pulse. In my case I have the timeout set to 10s. In the log snipped below, I get a pulse at 01:44:11 and another at 01:44:17, but at 01:44:21 (10 seconds after the penultimate pulse), I get the "No pulse detected for 10s message" and the sensor goes to 0, which is wrong:

The pulse meter I have setup to monitor wind speed doesn't seem to initially time out until after it's been triggered at least once, it will sit on unknown/NA status till it gets a single pulse, reports that value correctly, then times out as expected and reports 0

**Comment by timn** [comment]
I have the same issue. Accessing `.state` of the sensor gives nan. May I ask to rollback the latest change #4231 that apparently broke this? @TrentHouliston could you please take a look, or is there a way to fix forward?

The snippet that's breaking for me is this https://github.com/timn/esphome-meters/blob/main/gas/gas-meter-impuls.yaml.

I have confirmed that it works when downgrading to 2023.7.1:
```
docker run --rm -v "${PWD}":/config -it esphome/esphome:2023.7.1 run --device <ipaddr> <your_file.yaml>
```

**Comment by Grafton1046** [comment]
Work around I've found is to publish a value for the sensor on boot.

**Comment by TrentHouliston** [comment]
Hi, I'll take a look sorry I didn't see this earlier until I was pinged.

So by the looks, there are several issues that people are having in this thread?

> @egite  The pulse counter is going crazy/reporting unknown

If the reporting unknown is only at the start before any pulses have happened, that's likely the same issue others are discussing with timeouts not starting until the first pulse has been received.
As for the high pulse counts, do you have any other information or logs that could help me look into it?

> @Megu @slallemand @jmoery The sensor is reporting unknown

Is this on startup before a single pulse has been detected?

> @pmartyniuk resetting the total pulse count doesn't work?

This is surprising as I didn't even touch that code in my refactor, so it should still be working if it was working before. I'll look and see if I can see why it would have changed.

> @fabianschaefer The timeouts are not working as expected

So based on your data the penultimate pulse is where the timeout is being measured from rather than the last actual pulse?
I'll take a look but I'm struggling to see a mechanism in the code for that to happen.

> @Grafton1046 @timn  The pulse counter doesn't timeout on startup

What would be the preferred behaviour here?
Just send a 0 value on boot or timeout after the timeout on boot and send 0
I opted not to send 0 when starting up as I didn't know the state of the pulse meter and therefore have no information about the current pulse rate.
However, if the behaviour should be to timeout after `timeout` time has passed after startup without a pulse that can be fixed.

**Comment by TrentHouliston** [comment]
I have made a PR that will timeout the pulse meter on startup here
esphome/esphome#5388

**Comment by timn** [comment]
I would not want a 0 to be sent. This is a total increasing counter, so a state of 0 would give wrong results (I have seen this for other sensors in HA and this would cause the entire total to be counted for each period (e.g., day). This is why in the config I linked I restore the value from flash on boot. Do you have a spare board and sensor to test that config?

**Comment by TrentHouliston** [comment]
I was referring to publishing 0 for the pulse rate sensor rather than the total sensor. I agree for the total sensor that posting 0 on startup would cause problems, especially if your intention is to restore the value from flash.

What was the behaviour that it had previously for you with the total sensor?
As far as I can tell there should be no difference in behaviour for the output of the total sensor. The old code also just published a new update when it received a pulse and wouldn't have published anything prior to that.

**Comment by fabianschaefer** [comment]
> > @fabianschaefer The timeouts are not working as expected
> 
> So based on your data the penultimate pulse is where the timeout is being measured from rather than the last actual pulse? I'll take a look but I'm struggling to see a mechanism in the code for that to happen.

Yes that‘s correct. I‘ve now just doubled the timeout time in my config as a workaround, so it‘s not super urgent. Maybe the problem resolves itself when the root cause of some other issues here is fixed.

**Comment by Goodwillson** [comment]
I am getting also extreme sensor reads for wind speed after upgrading from 2023.5.5 to 2023.8.3. The wind speed goes up to 4000 km/h. I upped internal_filter to 100us, but I guess it may affect readings at higher wind speeds with such high limit. The extreme high wind speeds got filtered, but time to time I got speeds over 100km/h even at 100us. Is there any way to make the configuration compatible with new pulse meter?

- platform: pulse_meter
    pin:      
      number: GPIO25
      mode: INPUT_PULLUP
    id: wind_speed
    unit_of_measurement: 'm/s'
    name: "${friendly_name} wind speed"
    icon: 'mdi:weather-windy'
    internal_filter: 13us
    timeout: 5s
    filters:
      - multiply: 0.005560619
      - sliding_window_moving_average:
          window_size: 20
          send_every: 20

**Comment by OBerghmans** [comment]
I got similar results. Since 2023.8 I am measuring significantly more water usage via the config below. The water flow rate changed from around 10 liter/min to 130 liter/min (while my maximum possible in the house is 12 liter/min), and a daily usage from 150 liter to 1300 liter. I have to reset my Home Assistant utility meter for water every few days (subtracting about 3m³ each time).

I haven't played around with the parameters yet, but yesterday I compiled the 2023.7.1 firmware and flashed that to my board - as timn mentioned last week - as a temporary workaround.

```
  - platform: pulse_meter
    id: pulse_meter_water
    pin: ${ir_digital_pin}
    unit_of_measurement: "L/min"
    icon: "mdi:water"
    name: "Water Flow Rate"
    internal_filter: 150ms
    internal_filter_mode: PULSE
    accuracy_decimals: 3
    timeout: 30s
    total:
      name: "Water Usage"
      unit_of_measurement: "m³"
      icon: "mdi:water"
      device_class: water
      state_class: total_increasing
      accuracy_decimals: 3
      filters:
        - multiply: 0.001
```

**Comment by OBerghmans** [comment]
I still had big spikes with the new version (2023.9.1) so reverted again to 2023.7.1. I did run only a day and a half on the new version so I cannot say whether it is 'better' than previously, but after seeing a spikes to 25 and 100l/min I decided to roll back. I can test if needed, if someone has some pointers and test scenarios.

**Comment by TrentHouliston** [comment]
Hi @OBerghmans I have a branch which logs every interrupt that occurs which is useful for

**Comment by OBerghmans** [comment]
Thanks for the pointers @TrentHouliston ! I flashed the firmware, but I had to adjust the header file as I was getting compilation errors. A PR to your branch is created :-)

I will keep the logs open to see if I can get relevant interrupt data. Hopefully I get lucky because I have to keep it open in the browser (as my device is not connected via USB so I cannot check the serial connection). Or is there another way?

**Comment by TrentHouliston** [comment]
I've never tried this for longer periods myself, but if you have a computer you can run a command on and leave it you could run.
```sh
wget --user <web_user> --ask-password http://your_device.local/events
```
That just saves the log messages to a file, although the format isn't the nicest it's nothing a bit of cleaning can't fix

**Comment by TrentHouliston** [comment]
I know, the level of logging that branch does is excessive which is why I never actually put it into the main code even as a verbose log.
And yes the pulse filter is very good at that kind of noise since the state needs to be stable for filter_length before it counts.
If you do find a spike with logs that would be amazing to see.

**Comment by Goodwillson** [comment]
I included your logging component into my weather station, but ISR shows just zeros, despite I noticed about two or three huge spikes in wind speed. Filtered out whole log and there are only zeros logged everytime, but I see that pulse_meter component is taking long time to finish operation for some reason. I got that warning about 180x during the logging, but I noticed only about 2-3 spikes in wind data, so not sure if thats related.

```
id: 568830
event: log
data: [0;36m[D][pulse_meter:038]: ISR: 568824689 0 0 0 0 0[0m

id: 568836
event: log
data: [0;36m[D][sensor:094]: 'METEO  wind pulses total': Sending state 4714.00000 pulses with 0 decimals of accuracy[0m

event: state
data: {"id":"sensor-meteo__wind_pulses_total","value":4714,"state":"4714 pulses"}

id: 569006
event: log
data: [0;36m[D][pulse_meter:038]: ISR: 568996517 0 0 0 0 0[0m

id: 569011
event: log
data: [0;36m[D][pulse_meter:038]: ISR: 568996549 0 0 0 0 0[0m

id: 569024
event: log
data: [0;36m[D][sensor:094]: 'METEO  wind pulses total': Sending state 4716.00000 pulses with 0 decimals of accuracy[0m

event: state
data: {"id":"sensor-meteo__wind_pulses_total","value":4716,"state":"4716 pulses"}

id: 569033
event: log
data: [0;36m[D][sensor:094]: 'METEO wind speed': Sending state 558.67969 m/s with 2 decimals of accuracy[0m

id: 569033
event: log
data: [0;36m[D][sensor:094]: 'METEO wind speed': Sending state 558.67969 m/s with 2 decimals of accuracy[0m

id: 569049
event: log
data: [0;36m[D][sensor:094]: 'METEO wind speed (km/h)': Sending state 2011.24683 km/h with 2 decimals of accuracy[0m

id: 569052
event: log
data: [0;36m[D][sensor:094]: 'METEO wind gust 60s': Sending state 2011.24683 km/h with 1 decimals of accuracy[0m

event: state
data: {"id":"sensor-meteo_wind_gust_60s","value":2011.247,"state":"2011.2 km/h"}

id: 569069
event: log
data: [0;36m[D][sensor:094]: 'METEO wind gust 10min': Sending state 2011.24683 km/h with 1 decimals of accuracy[0m

event: state
data: {"id":"sensor-meteo_wind_gust_10min","value":2011.247,"state":"2011.2 km/h"}

event: state
data: {"id":"sensor-meteo_wind_gust_10min","value":2011.247,"state":"2011.2 km/h"}

event: state
data: {"id":"sensor-meteo_wind_speed_kmh","value":2011.247,"state":"2011.25 km/h"}

event: state
data: {"id":"sensor-meteo_wind_speed","value":558.6797,"state":"558.68 m/s"}

id: 569093
event: log
data: [0;33m[W][component:204]: Component pulse_meter.sensor took a long time for an operation (0.09 s).[0m

id: 569105
event: log
data: [0;33m[W][component:205]: Components should block for at most 20-30ms.[0m

id: 569304
event: log
data: [0;36m[D][sensor:094]: 'METEO wind speed average': Sending state 281.53098 m/s with 2 decimals of accuracy[0m

id: 569319
event: log
data: [0;36m[D][sensor:094]: 'METEO wind speed average (km/h)': Sending state 1013.51147 km/h with 2 decimals of accuracy[0m

```

My setup was following:

- platform: pulse_meter
    pin:      
      number: GPIO25
      mode: INPUT
    id: wind_speed
    unit_of_measurement: 'm/s'
    name: "${friendly_name} wind speed"
    icon: 'mdi:weather-windy'
    internal_filter: 20us
    timeout: 5s
    filters:
      - multiply: 0.005560619
      - sliding_window_moving_average:
          window_size: 20
          send_every: 20
    total:
      name: "${friendly_name}  wind pulses total"

**Comment by TrentHouliston** [comment]
That's really odd, any time a value is published (that isn't a timeout) it should have a corresponding ISR state.
It's possible that the pulse rate is too high and the ESP can't keep up with the logging.

Also all 0s in the ISR implies that both the current and previous value detected in the ISR (that's supposed to fire on change) are 0. So changing from 0 to 0... The only time I've seen that happen is when the signal is really noisy, and even then the pulse filter is coded to treat that kind of same value noise as a reset which should lower the rate

**Comment by OBerghmans** [comment]
I finally had some good readings this evening. From what I see in my logs I might just need to increase the internal_filter, but I will link the logs so maybe you can determine something else:
- https://esphomepulsemeterisr.s3.eu-west-1.amazonaws.com/esp32s-basement-water.log
- https://esphomepulsemeterisr.s3.eu-west-1.amazonaws.com/ISR.xlsx - This is the logs lines filtered, exported to CSV and imported in Excel. I also included a screenshot of my home assistant graph of the corresponding timeframe in this excel file. During this frame no water was used in the house.

Hopefully that information can help, and I really appreciate your help with this!

**Comment by elpirato83** [comment]
> ### The problem
> Greetings.
> 
> Upgrading from 2023.7.1 to 8.0, the pulse_meter function seems to have broken/changed. The below now causes the counter to go crazy and increment at a high rate endlessly. Removing the internal_filter causes the initial count after reset to be a randomly high number. The pulse count afterwards counts, though is several orders of magnitude higher than actual. Removing the debounce filters also causes endless increments.
> 
> I have a separate ESP8266 on an identical but separate water meter running the same code that no longer reports its count. The sensor shows as "unknown".
> 
> Please let me know how I can help.
> 
> ### Which version of ESPHome has the issue?
> 2023.8.0
> 
> ### What type of installation are you using?
> Home Assistant Add-on
> 
> ### Which version of Home Assistant has the issue?
> 2023.8.2
> 
> ### What platform are you using?
> ESP8266
> 
> ### Board
> _No response_
> 
> ### Component causing the issue
> _No response_
> 
> ### Example YAML snippet
> ```yaml
> sensor:    
>   - platform: pulse_meter
>     pin: 
>       number: GPIO5  #D1
>       mode:  
>         input: true
>     name: "Pulse Meter"
>     unit_of_measurement: "gal/min"
>     internal_filter: 50ms   
>     internal_filter_mode: EDGE
>     accuracy_decimals: 1
>     timeout: 10sec
>     icon: "mdi:gauge"    
>     filters: 
>      - multiply: 0.0748052    
>      - debounce: 50ms      
>     total:
>       name: "Total"
>       accuracy_decimals: 3
>       unit_of_measurement: "gal"
>       device_class: water
>       state_class: total_increasing
>       filters: 
>         - multiply: 0.0748052
>         - debounce: 50ms
> ```
> 
> ### Anything in the logs that might be useful for us?
> _No response_
> 
> ### Additional information
> _No response_

I had exactly the same issue on my esp32 water meter board.
I had to add the code below to prevent the pulse meter going crazy. After upgrading to 2023.8 the Esp32 baord was picking up a lot of noise on the input pin which caused this. 

pin:
  mode:
    input: true
    pullup: true

**Comment by CestLaGalere** [comment]
@TrentHouliston - are you looking at this at all please? Your August refactor seems to have caused a number of issues which are still outstanding.  Thanks

**Comment by TrentHouliston** [comment]
I am still looking into this as I get time however as I don't seem to get any of the issues described It's very hard for me to work out what is wrong.

From what I'm seeing it looks like it's the result of very noisy data which is what the filter is for and I'm trying to work out the mechanism that this noise is slipping through.
The code that does the pulse filtering isn't that long so there aren't that many places something could happen.
I'm wondering if some other change somewhere else changed something which is showing up as errors here, or if it's something to do with this new filtering code.

```cpp
  // Get the current time before we do anything else so the measurements are consistent
  const uint32_t now = micros();
  const bool pin_val = sensor->isr_pin_.digital_read();

  // A pulse occurred faster than we can detect
  if (sensor->last_pin_val_ == pin_val) {
    // If we haven't reached the filter length yet we need to reset our last_intr_ to now
    // otherwise we can consider this noise as the "pulse" was certainly less than filter_us_
    if (now - sensor->last_intr_ < sensor->filter_us_) {
      sensor->last_intr_ = now;
    }
  } else {
    // Check if the last interrupt was long enough in the past
    if (now - sensor->last_intr_ > sensor->filter_us_) {
      // High pulse of filter length now falling (therefore last_intr_ was the rising edge)
      if (!sensor->in_pulse_ && sensor->last_pin_val_) {
        sensor->last_edge_candidate_us_ = sensor->last_intr_;
        sensor->in_pulse_ = true;
      }
      // Low pulse of filter length now rising (therefore last_intr_ was the falling edge)
      else if (sensor->in_pulse_ && !sensor->last_pin_val_) {
        sensor->set_->last_detected_edge_us_ = sensor->last_edge_candidate_us_;
        sensor->set_->count_++;
        sensor->in_pulse_ = false;
      }
    }

    sensor->last_intr_ = now;
    sensor->last_pin_val_ = pin_val;
  }
```

**Comment by fabianschaefer** [comment]
@TrentHouliston, if you have issues reproducing the issue, maybe it could be a start to look into the timeout issue that I mentioned above (https://github.com/esphome/issues/issues/4807#issuecomment-1685134880)
This one is very easily reproducible and maybe the two issues are linked. So if you solve the root cause of the timeout issue, the other issue might be solved as well.

**Comment by CestLaGalere** [comment]
@TrentHouliston thanks - I am not sure it is just to do with the noisy data issue - I have a pulse monitor attached to a photo diode sensor board (as an example [https://ardubotics.eu/en/sensors/1164-lm393-photodiode-sensor-module.html](https://ardubotics.eu/en/sensors/1164-lm393-photodiode-sensor-module.html) - Whilst I have not put an oscilloscope on the output as it is electronically generated it ought to be fairly clean!
I have just rolled it back and can confirm that 2023.7.1 gives the correct output (in my case I am measuring light pulses from an electricity meter) and have a ct clamp at the moment which I can compare it to.  Moving to 2023.8.0 or later the reading goes from about .5kW up to 170kW). The meter gives one light flash every second or two, so pulses of low frequency. If there is any debug info I can provide that might help please let me know.

**Comment by felixstorm** [comment]
@TrentHouliston First of all, thanks for taking the time and effort to clean up the old `pulse_meter` code. I am trying to understand your new code and potentially help out with at least some of the issues mentioned here but need to be sure on one question first: In mode `PULSE`, should a pulse be regarded as completed on a rising edge or on a falling edge, i.e. are we counting low pulses (completed when rising again) or high pulses (completed when falling again)?

The docs do not seem to really be precise on the expected behavior but [in one place talk about a "high pulse"](https://github.com/esphome/esphome-docs/blob/359f3595fa0a2a28f5065629d113d62ac5023fba/components/sensor/pulse_meter.rst?plain=1#L39). But [in your current code](https://github.com/esphome/esphome/blob/27b593ba85d26be8445b7868dc56db51304e198b/esphome/components/pulse_meter/pulse_meter_sensor.cpp#L134) the pulse gets counted on a _rising_ edge, i.e. it looks like we would actually count _low_ pulses (completed when signal is rising again) - that's why I am asking.

Also, when looking at the history of your PR, it seems like counting had been switched from falling edge to rising edge at some point on the way without that fact being mentioned in the comments thread: https://github.com/esphome/esphome/pull/4231/commits/80e565f2abb59df28018ef5283f89e6fe13c2fb1#diff-2421f768261b150494a9645d18fee16f9463486eb5c1f849b9f8f7ab8c96e283L107-R115

**Comment by TrentHouliston** [comment]
Hi @felixstorm that is one of the parts when writing the code that caused some confusion for me. I tried to comment on it but I guess it wasn't clear enough.

The reason that the pulse gets counted on the rising edge is that iss the first time that you can know that there has been a low signal of sufficient length.

The pulse starts on a rising edge and then is completed on the following falling edge, however  until we get another rising edge where we can see that the value was low for at least `filter_width` we can't say that the previous high to low was valid.

It's for filtering out the noise case where say you have a rising edge, then stay high for filter width, then have a very short low then high (noise), in that case since the low pulse wasn't filter_width it doesn't count as the "falling edge"

**Comment by felixstorm** [comment]
@TrentHouliston Thanks for your reply - that makes things a lot clearer for me.

IMHO the delay caused by having to wait until the next rising edge is something that‘s not desirable at all for sources with lower pulse count (e.g. water flow based on 1-liter-pulses from a traditional water meter) as it means that in reality the pulse_meter will always show the flow one interval late - in fact that was the exact reason why I started looking at the code details in the first place.  
But in contrary to the delay_on_off filter (which can act right after a delay has passed without having to wait for another pulse at all) the current ISR-based solution only can act if the ISR gets triggered by another edge (which might be very late). As you say it might be worth trying to push that candidate to the main loop immediately and let the loop process it as soon as enough time has passed since the pulse has fallen without rising again…

Looking at the different reports here in detail it seems that most of the people with trouble are (implicitly) using `EDGE` mode (besides @OBerghmans - but if I am not mistaken he mentions that he should be able to fix his setup by incrementing the filter duration) - which uses a different ISR code path. I also tried edge mode and noticed unreal high results but immediately switched to pulse mode. I can try to debug into edge mode again a bit more, but maybe `PULSE` mode should become the default (after potentially changing it to count pulses sooner as discussed above) since it now is the better and more error-proof setting?

I will think about it a bit more and see if I can come up with some other helpful idea.

**Comment by felixstorm** [comment]
About ISR logging: In my tests I utilized a FreeRTOS queue for ISR logging and I should be able to implement it in a way that should have almost zero impact outside of `VERBOSE` or `VERY_VERBOSE` logging. I might bring this up as a PR for permanent inclusion in main line code so people can debug their setup more easily (without impacting other „normal“ users).

**Comment by felixstorm** [comment]
One more thing: `PULSE` mode often seems to discard the very first pulse after a restart (which messes up the counter). Setting `last_pin_val_` to the opposite of the current pin state in `setup()` fixed that for me - I could bring this up as a PR as well…

**Comment by TrentHouliston** [comment]
> About ISR logging: In my tests I utilized a FreeRTOS queue for ISR logging and I should be able to implement it in a way that should have almost zero impact outside of VERBOSE or VERY_VERBOSE logging. I might bring this up as a PR for permanent inclusion in main line code so people can debug their setup more easily (without impacting other „normal“ users).

If you can find a way to include it without making the code much more complicated that would be great. Then it would be easy to just turn it on with very verbose logging. The main reason I didn't include it originally was that I wanted to keep the ISRs as minimal as possible as how long those ISRs take to run directly impacts how well the filter can operate. If they take too long and pulses/noise start building up faster than the ISR can process that's when you can start getting weird behaviour.

> One more thing: PULSE mode often seems to discard the very first pulse after a restart (which messes up the counter). Setting last_pin_val_ to the opposite of the current pin state in setup() fixed that for me - I could bring this up as a PR as well…

That might be a good idea, it should help a lot of the time if the signal is clean enough. Make sure you set it as close as possible to where the ISR itself gets attached to limit the chance of the pin changing

Also if you are going to look at the pulse filter itself how I started was just listing all the options like in the code below and working out what to do in each, then reducing the logic down after.

<details>
  <summary>Code</summary>

```cpp
  // This is an interrupt handler - we can't call any virtual method from this method
  // Get the current time before we do anything else so the measurements are consistent
  const uint32_t now = micros();
  const bool pin_val = sensor->isr_pin_.digital_read();

  const bool length_y = now - sensor->last_intr_ >= sensor->filter_us_;
  const bool length_n = !length_y;
  const bool pulse_y = sensor->in_pulse_;
  const bool pulse_n = !pulse_n;
  const bool change_y = sensor->last_pin_val_ != pin_val;
  const bool change_n = !change_y;
  const bool hi = pin_val;
  const bool lo = !pin_val;

  if (false) {
  } else if (pulse_y && len_y && change_y && hi) { // 1 1 1 1
  } else if (pulse_y && len_y && change_y && lo) { // 1 1 1 0
  } else if (pulse_y && len_y && change_n && hi) { // 1 1 0 1
  } else if (pulse_y && len_y && change_n && lo) { // 1 1 0 0
  } else if (pulse_y && len_n && change_y && hi) { // 1 0 1 1
  } else if (pulse_y && len_n && change_y && lo) { // 1 0 1 0
  } else if (pulse_y && len_n && change_n && hi) { // 1 0 0 1
  } else if (pulse_y && len_n && change_n && lo) { // 1 0 0 0
  } else if (pulse_n && len_y && change_y && hi) { // 0 1 1 1
  } else if (pulse_n && len_y && change_y && lo) { // 0 1 1 0
  } else if (pulse_n && len_y && change_n && hi) { // 0 1 0 1
  } else if (pulse_n && len_y && change_n && lo) { // 0 1 0 0
  } else if (pulse_n && len_n && change_y && hi) { // 0 0 1 1
  } else if (pulse_n && len_n && change_y && lo) { // 0 0 1 0
  } else if (pulse_n && len_n && change_n && hi) { // 0 0 0 1
  } else if (pulse_n && len_n && change_n && lo) { // 0 0 0 0
  }
```

</details>

**Comment by OBerghmans** [comment]
To follow up: I am indeed using PULSE mode. And I did increase the filter duration. That did not completely solve it. I just don't have any high water flow rate readings anymore (max 10 liter per minute), but it just keeps going up slowly in certain cases. For example: if I would believe the sensor, then we would have used 200l water during the night while everyone was in bed 😏 

So I am still following along and will also check the code. Feel free to add me in PRs for reviewing.

**Comment by OBerghmans** [comment]
Just as some clarification and/or proposal to change the 3-deep nested if structure, I rewrote the pulse_intr method just for my understanding. _So anyone reading this, this is my own snippet and not the real code:_

```
    // Get the current time and pin value
    const uint32_t now = micros();
    const bool pin_val = sensor->isr_pin_.digital_read();
    
    // If the pin value hasn't changed and the pulse is too short, treat as noise and reset last_intr_
    if (sensor->last_pin_val_ == pin_val) {
        if (now - sensor->last_intr_ < sensor->filter_us_) {
            sensor->last_intr_ = now;
        }
        return; // Early exit for noise or a very fast pulse
    }
    
    // Handle rising and falling edges
    if (now - sensor->last_intr_ > sensor->filter_us_) {
        if (!sensor->in_pulse_ && sensor->last_pin_val_) {
            // Detected rising edge
            sensor->last_edge_candidate_us_ = sensor->last_intr_;
            sensor->in_pulse_ = true;
        } 
        else if (sensor->in_pulse_ && !sensor->last_pin_val_) {
            // Detected falling edge
            sensor->set_->last_detected_edge_us_ = sensor->last_edge_candidate_us_;
            sensor->set_->count_++;
            sensor->in_pulse_ = false;
        }
    }

    // Update last interrupt time and last pin value for changes
    sensor->last_intr_ = now;
    sensor->last_pin_val_ = pin_val;
```

@TrentHouliston Does this behave exactly the same or did I make a wrong assumption? The comments in the real code (also) did confuse me on what was considered the rising and falling edge.
I did exit early in case the last_pin_val was the same as pin_val just so the remainder of the function is not as nested as it was.

**Comment by TrentHouliston** [comment]
https://godbolt.org/z/Yrq7zKzMr

Looks like it generates the exact same assembly, so the behaviour would be identical

**Comment by CestLaGalere** [comment]
I can confirm that changing the default setting in my case to
```
internal_filter_mode: PULSE
```
gives the expected output - I can try ang get a trace of the inout signal if that would help - the default value of EDGE breaks with the refactoring.

**Comment by TrentHouliston** [comment]
> And reading through your last paragraph, you probably meant to says the same thing with it 😔 - correct?

Correct, I realised as I was reading the code there is no need for it. I should have written it to publish the moment I saw the falling edge. It had gone thorough a few iterations by that point and I must have missed that was an option.

@OBerghmans Yeah godbolt is really useful for trying to work out what code is doing. Helped me a lot when learning what types of things the compiler can optimise.
At the end of the day if the assembly is the same it all comes down to readability and personal preference.

**Comment by OBerghmans** [comment]
@felixstorm Do you have your FreeRTOS queue code pushed somewhere? I'm interested so I could include it in my branch and further test my case.

**Comment by felixstorm** [comment]
I have not pushed it yet since I had been playing around also with the filter, but I should hopefully be able to push it tomorrow during the day (CET). It is not magic but a bit tricky since only ESP8266 and ESP32 are based on FreeRTOS but not the other cores supported by ESPHome and therefore it should only be enabled if FreeRTOS is available to avoid compilation errors on other platforms.

BTW - the ISRs on any edge did not seem to be really stable during my tests on an ESP32: I had situations where the pin never read 1 after an ISR and therefore no pulse got detected even though a button had been pressed for several seconds. It might have been that the signal was very noisy and the pin had gone back to 0 for a short moment again but since we can only sample the pin state right in the ISR with the current concept I have no real idea how to deal with those situations. Also it does not seem possible to attach separate ISRs at the same time to rising and falling edges either (to get the pin state based on the specific ISR that got called) :-(

**Comment by felixstorm** [comment]
@TrentHouliston @OBerghmans Ok, I just now found some time to push my work-in-progress with ISR debug logging (and initial setting of `last_pin_val_` to not miss the first pulse): https://github.com/felixstorm/esphome/commits/dev-felixstorm

I will probably not be able to do much further work on this during the next few days, so please feel free to copy the code for a PR etc.

**Comment by TrentHouliston** [comment]
@fabianschaefer Just FYI with the help of https://github.com/esphome/esphome/pull/5813 I've realised what is likely causing your timeouts from the penultimate pulse problem. As I described in that PR I think there is a solution that I'll either try to find time to implement sometime or help with that PR to implement the fix.

**Comment by petewill** [comment]
For the last couple of weeks, I have been reading through this and other issues/posts trying to find a fix for my pulse meter.  I have a water meter (DAE VM-75P) that uses a dry contact sensor to send a pulse.  The problem is the signal is noisy and the contact sensor stays high half of the time and low the other half.  So, it's high 0-0.5 gal and low 0.5-1 gal.  This means it could be high or low for hours if we're not using water.  All that to say it doesn't pulse quickly even when we're using water.  If we are using the maximum amount of water we could, the fastest it would pulse is about once every 2 seconds.  I first tried setting the internal_filter_mode to Edge and that resulted in many more pulses than there actually were.  When I set it to Pulse it doesn't pick up all my pulses.  It lags behind about 9 pulses every 100 pulses.  I tried with @hannesb's code mentioned in this issue https://github.com/esphome/esphome/pull/5813#issuecomment-1828743336 and I am getting similar results.  Here is my code:

```
sensor:
  - platform: pulse_meter
    pin:
      number: 33
      inverted: true
      mode:
        input: true
        pullup: true
    name: "Water Usage Rate"
    id: water_meter
    device_class: water
    unit_of_measurement: "gal/min"
    icon: "mdi:water"
    internal_filter: 200ms
    internal_filter_mode: pulse
    timeout: 30 min
    accuracy_decimals: 4
    total:
      name: "Water Meter Total"
      unit_of_measurement: "gal"
      id: water_total
      device_class: water
      state_class: total_increasing
      accuracy_decimals: 0
      filters:
        - lambda: |-
            id(total_pulses) = id(water_total).state;
            //ESP_LOGI("main", "total_pulses value: %f", id(total_pulses));
            id(flash_led).turn_on();
            return x;
```

I have tried adjusting the "internal_filter" without any luck.

For reference, I have set up a test binary sensor and it worked flawlessly for the couple of days that I tested it.  Here was my code for that:

```
binary_sensor:
  - platform: gpio
    pin:
      number: 33
      inverted: true
      mode:
        input: true
        pullup: true
    id: water_counter
    internal: true
    on_press:
      then:
        - lambda: |-
            id(total_pulses) += 1;
            ESP_LOGI("main", "total_pulses value: %f", id(total_pulses));
            id(flash_led).turn_on();
    filters:
      - delayed_on: 200ms
      - delayed_off: 200ms
```
Does anyone have any ideas for how I can improve my results?  I tried a few external debounce methods (add a capacitor and resistors of various sizes) but that didn't help either.

**Comment by TrentHouliston** [comment]
I made a PR about 3 weeks ago to fix the timeout issue mentioned by https://github.com/esphome/esphome/pull/5813 that would help with some of the timeout issues that occur which might make it miss pulses events if they're quite slow.

Nobody has had a look at it yet and I have difficulty testing it because none of my sensors I have access to really warrant a pulse filter currently.

Although the total should still be correct even if due to timeouts there are less actual events so the fact you're missing 9/100 is still concerning.
The only possible case I've seen implied by the other issues I've seen over time is that sometimes if the pluses change faster than the ISR it seems that it gets into weird states.

I just had an idea for something that might help so I'll make a quick branch off 5813 so you can test it if you'd like.

**Comment by petewill** [comment]
Thanks @TrentHouliston!  I just uploaded the code with your loop_pulse_isr branch.  I'm new to this, so I just want to make sure this is the correct way to do that:

```
external_components:
  - source: 
      type: git
      url: https://github.com/TrentHouliston/esphome
      ref: loop_pulse_isr
    components: [ pulse_meter ]
```

If I missed something, please let me know.  Otherwise, I'll report back on how things went soon.

**Comment by TrentHouliston** [comment]
That is the verbose way and will work, you can also just do
```yaml
external_components:
  - source: github://TrentHouliston/esphome@loop_pulse_isr
    components: [ pulse_meter ]
```

**Comment by petewill** [comment]
Thanks Trent.  Unfortunately the updated code didn't seem to work.  I updated the esp32 about 2 days ago.  The reading on the meter was 3423. I thought things were working well but over the last few hours it has gotten worse.  It was only off by a gallon as of last night but this morning it's off by 5.  The meter reading is now 3512 but the pulse meter reading is 3507.  I'm not sure why it would be missing pulses.  So strange.

**Comment by TrentHouliston** [comment]
I've got another one to try if you want, it doesn't use an ISR at all and just reads the pin value each loop. So it won't work at all if you have very fast pulses, but it sounds like your pulse rate is quite slow (compared to the speed of the loop)

https://github.com/TrentHouliston/esphome/tree/no_interrupt

**Comment by petewill** [comment]
Thanks.  I just uploaded the code. I'll let you know how it goes.  And, yes, my pulse rate is very slow compared to the loop.

**Comment by petewill** [comment]
@TrentHouliston That seems to have worked!  We have used around 180 gallons and it is reading out exactly the same as the water meter.  I'll keep monitoring it and update in a few days.  If this fix works, is it possible to add in a Configuration Variable to the Pulse Meter that would allow us to bypass the ISR like you have done with this test code?

**Comment by petewill** [comment]
Well, that seems to have fixed it.  The esp32 reading is still the same as the water meter after another 3 days and approximately 150 gallons.

**Comment by TrentHouliston** [comment]
Fixed for you but broken for anyone with higher rates! I'll try have a think to see if I can find a way to make the loop correct when it misses interrupts without making the code horrible. If not then yes an option that doesn't use interrupts could be a potential solution for slower pulse rates. 

Although the better option would be to know why the esp is missing interrupts in the first place.

**Comment by hannesb** [comment]
Their are a lot reports with interrupt problems with ESP32, i.e.:
https://github.com/espressif/arduino-esp32/issues/4172
https://github.com/espressif/arduino-esp32/issues/1111
My conclusion is: the ESP32 has not build in signal conditioning in hardware. So if you are using interrupts with ESP32, you need some extra hardware for signal conditioning. RC filtering does not work, I recommend a Schmitt Trigger setup.

**Comment by TrentHouliston** [comment]
@hannesb That's very useful information and explains a lot of issues people have described.
The PR I made below should hopefully address that specific issue.

@petewill This pull request has a better fix https://github.com/esphome/esphome/pull/6126
If you could give it a go and it works that'd be amazing since then it should work both with interrupts and when they're missed

**Comment by petewill** [comment]
@TrentHouliston Thank you.  I just uploaded the code.  I used this branch: https://github.com/TrentHouliston/esphome/tree/loop_isr_fix
It looks like that's what you used for the pull request but I don't use git very often, so I just want to make sure that is correct.

**Comment by TrentHouliston** [comment]
That's corect, for PRs though in the future you can just use
```yaml
  - source: github://pr#6126
    components: [ pulse_meter ]
```

**Comment by petewill** [comment]
@TrentHouliston Things are looking good so far.  We have used around 400 gallons in the last week and the count is only off by one gallon (the esp is reading 1 more).  I can definitely live with that.  Thanks for all your work on this!

**Comment by Qhilm** [comment]
I am not sure it's related, but I have the opposite issue, the `pulse_meter` total is adding a pulse from time to time. It appears it is being added either at the beginning or at the end of a pulse series.

But since I have `internal_filter: 3400ms`, I don't see where these additional pulses could be coming from.

I have connected the same sensor (a reed switch) to a second GPIO and I am reading the input using `binary_sensor` and this [external component](github://syssi/esphome-total-count@main) for counting the total nb of pulses. The `binary_sensor` total is absolutely rock solid so far, not a single pulse deviation in over 10000 pulses, comparing it to the meter (it's a wheel with a magnet). Hence I think the issue appears to be with `pulse_meter` for me, but I do not understand how it could be **adding** pulses.

Yaml and details in [this HA community thread](https://community.home-assistant.io/t/pulse-meter-counting-too-many-pulses/702328).

**Comment by TrentHouliston** [comment]
@Qhilm Hi, Have you tried my pull request above? Unfortunately it seems that none of the maintainers of esphome have had time to review it but it'll likely solve the issues you're seeing

**Comment by Qhilm** [comment]
So I switched to `internal_filter_mode: PULSE` and I stopped having too many pulses and started having too few.

Then I did what you recommended @TrentHouliston and added:

```yaml
external_components:
  - source: github://pr#6126
    components: [ pulse_meter ]
```

I does make a difference for sure, but now it seems I have too many pulses again, compared to the binary sensor total (and compared to what I actually read on the meter). Not a lot, a couple per day, I would say one in a thousand, but it's completely inexplicable for me, since the binary sensor connected to the exact same reed switch is counting correctly. I see things like this for example, pulse_meter total (in red) is jumping by 2 on rising edge (binary signal in blue, binary sensor total in orange):

<img width="459" alt="image" src="https://github.com/esphome/issues/assets/3350433/440c228d-ec40-40a4-a0c4-07587a98d783">

Do you think it's possible for the interrupt to cause a pulse to be counted twice, with your PR?

**Comment by LucasCZE** [comment]
I also sometimes (about once per 2 days) observe the total count jumping by 2 on a rising edge. I am on the version 2024.10.2 so I don't think it is related to the changes in the PR 6126 (as I don't have these changes but I see the same issue).

You can see the issue in the graphs below. When the issue occurred, the volume flow rate was set to 2,000,000 m³/h (in this case). That would mean 55,555.55 revolutions per second, which would mean that one revolution took 18 μs (HIGH for 2.73 μs and LOW for 15.27 μs). What's interesting is that I have the filter set to 500 ms, so that kind of volume flow rate should be filtered out with a very large margin. With the filter set to 500 ms, it should not be possible to achieve a volume flow rate greater than 36 m³/h.

Personally, I think the issue might be somewhere in the code for early counting the unprocessed edges. We early count the edge in the loop and then we count the same edge again in the next loop by mistake. But the issue will only appear if the interruption occurs at a certain moment. That would explain why it only happens once every few days. But I don't know much about embedded programming and I don't see any problem in this code, so maybe I'm completely wrong.

What I do know is that the incorrect incrementation of the total count will not occur in one loop, but in two following loops. If this happened in one loop, the volume flow rate would not be changed because it would just set `meter_state_` from `TIMED_OUT` to `RUNNING`. Also, I see that the total count first published state 4.10 m³ and than published state 4.11 m³ only after a little moment. That would not be possible in one loop. In one loop, state 4.11 m³ would be published directly.

![image](https://github.com/user-attachments/assets/7bfc1ad2-c550-4611-bda3-f3b63229f092)

![image](https://github.com/user-attachments/assets/5865e655-5910-4664-b630-ceede31d7661)

```
sensor:
  - platform: pulse_meter
    id: gas_volume_flow_rate_sensor
    name: "Utility Meter - Gas - Volume Flow Rate"
    unit_of_measurement: "m³/h"
    device_class: volume_flow_rate
    state_class: measurement
    icon: "mdi:fire"
    accuracy_decimals: 2
    filters:
      - multiply: 0.6
    pin:
      number: GPIO22
      inverted: true
      mode: INPUT
    internal_filter_mode: PULSE
    internal_filter: 500ms
    timeout: 2min
    total:
      id: gas_total_consumption_sensor
      name: "Utility Meter - Gas - Total Consumption"
      unit_of_measurement: "m³"
      device_class: gas
      state_class: total_increasing
      icon: "mdi:meter-gas-outline"
      accuracy_decimals: 2
      filters:
        - multiply: 0.01
```

**Comment by TrentHouliston** [comment]
This is off the PR https://github.com/esphome/esphome/pull/6126 correct?

**Comment by boarder876** [comment]
I don't know since which version I have problems with unexpected high values, but even with the filter the values are not limited anymore.
I've downgraded to 2023.7 and everything is again fine with the same configuration.

The highest value was like 2Mio Pulses per Minute with an internal_filter: 400ms

**Comment by LucasCZE** [comment]
The issue is still present in 2024.11.3

![image](https://github.com/user-attachments/assets/580c4bb3-0644-4858-ad05-6cb3045dc92e)

I just started using the Pulse Meter version before the code for early counting the unprocessed edges was added (added in 2024.5.0). At least I hope the code below achieves that. I'll let you know if the behavior changes or not.

```
external_components:
  - source: github://esphome/esphome@2024.4.2
    components: [pulse_meter]
```

**Comment by LucasCZE** [comment]
Version 2024.4.2 has been running for almost 5 days and the issue has not recurred yet. I think if there was the issue present, it would have shown up by now. So it looks like the issue must have been caused by the changes in [this PR](https://github.com/esphome/esphome/pull/6014).

**Comment by LucasCZE** [comment]
2024.4.2 has been running for 13 days without any issue.

![image](https://github.com/user-attachments/assets/6179202e-4d0e-49e2-a641-25be028dbc8d)

I am going to try 2024.12.2 where new ESP-IDF 5.1.5 is used. Maybe that will bring some improvement.

**Comment by LucasCZE** [comment]
I didn't realize that I'm using the Arduino framework, not ESP-IDF. So, of course, version 2024.12.2 didn't change anything.

![image](https://github.com/user-attachments/assets/2a7c9d42-1d45-4e9c-be6c-537ab43b1cc5)

I am rolling back to the Pulse Meter 2024.4.2 + everything else from the latest 2024.12.2.

**Comment by jires** [comment]
Anyone tried if issue persist in 2025.2.0?
I have the same problem with my gas meter and running now in version 2023.7.1 to avoid this, but ESP is restarting from time to time. Restarting problems has been fixed in later releases (don't remember which one), so is pretty annoying.

**Comment by TrentHouliston** [comment]
As far as I can tell there haven't really been any code changes to the pulse meter except for one which added more logging.

I doubt that anything has changed. I'm still waiting for PRs I made to fix bugs in pulse meter to get in from over a year ago https://github.com/esphome/esphome/pull/6126

It's somewhat demotivating to try and fix things when I know that any fix I do will take months or years to get into the codebase. I've mostly just been fixing things that bother me on my own systems and using my own forks.

**Comment by LucasCZE** [comment]
I just updated the pulse_meter to the latest version (2025.6.3). That includes the changes mentioned by TrentHouliston in the previous post. Since the bug only appears once every few days, we have to wait for the result. I will write the result here later.

**Comment by LucasCZE** [comment]
I encountered the issue again. This time I have the logs.

I think the only interesting entry in the log is this one:
`09:52:18	[V][pulse_meter:096]: New pulse, delta: 66 µs, count: 1, width: 66.00000 µs`

The internal_filter is set to 500ms. So it seems that under certain circumstances the filter does not work.

<img width="2141" height="807" alt="Image" src="https://github.com/user-attachments/assets/aec25d45-8610-45f5-ba7c-d7c4e3709a64" />

```
09:50:23	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 0.83007 L/min with 1 decimals of accuracy
09:50:23	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 0.830073
09:50:23	[V][pulse_meter:096]: New pulse, delta: 72282759 µs, count: 1, width: 72282760.00000 µs
09:50:23	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 253.00000 L with 0 decimals of accuracy
09:50:23	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 253.000000
09:50:33	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 0.70262 m³/h with 2 decimals of accuracy
09:50:33	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 1.171031
09:50:33	[V][pulse_meter:096]: New pulse, delta: 51236894 µs, count: 1, width: 51236896.00000 µs
09:50:33	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.12000 m³ with 2 decimals of accuracy
09:50:33	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 112.000000
09:50:37	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 4.042431
09:50:37	[V][pulse_meter:096]: New pulse, delta: 14842554 µs, count: 1, width: 14842554.00000 µs
09:50:37	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 254.00000 L with 0 decimals of accuracy
09:50:37	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 254.000000
09:50:38	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 4.04243 L/min with 1 decimals of accuracy
09:50:47	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 6.56894 L/min with 1 decimals of accuracy
09:50:47	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 6.568938
09:50:47	[V][pulse_meter:096]: New pulse, delta: 9133896 µs, count: 1, width: 9133896.00000 µs
09:50:47	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 255.00000 L with 0 decimals of accuracy
09:50:47	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 255.000000
09:50:57	[D][sensor:099]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Sending state -54.00000 dBm with 0 decimals of accuracy
09:50:57	[V][sensor:045]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Received new state -54.000000
09:51:12	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 0.90947 m³/h with 2 decimals of accuracy
09:51:12	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 1.515785
09:51:12	[V][pulse_meter:096]: New pulse, delta: 39583440 µs, count: 1, width: 39583440.00000 µs
09:51:12	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.13000 m³ with 2 decimals of accuracy
09:51:12	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 113.000000
09:51:13	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 2.30841 L/min with 1 decimals of accuracy
09:51:13	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 2.308411
09:51:13	[V][pulse_meter:096]: New pulse, delta: 25991902 µs, count: 1, width: 25991902.00000 µs
09:51:13	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 256.00000 L with 0 decimals of accuracy
09:51:13	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 256.000000
09:51:27	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 4.03944 L/min with 1 decimals of accuracy
09:51:27	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 4.039436
09:51:27	[V][pulse_meter:096]: New pulse, delta: 14853558 µs, count: 1, width: 14853558.00000 µs
09:51:27	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 257.00000 L with 0 decimals of accuracy
09:51:27	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 257.000000
09:51:38	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 5.52656 L/min with 1 decimals of accuracy
09:51:38	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 5.526563
09:51:38	[V][pulse_meter:096]: New pulse, delta: 10856658 µs, count: 1, width: 10856658.00000 µs
09:51:38	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 258.00000 L with 0 decimals of accuracy
09:51:38	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 258.000000
09:51:42	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 14.76804 L/min with 1 decimals of accuracy
09:51:42	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 14.768042
09:51:42	[V][pulse_meter:096]: New pulse, delta: 4062827 µs, count: 1, width: 4062827.00000 µs
09:51:42	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 259.00000 L with 0 decimals of accuracy
09:51:42	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 259.000000
09:51:46	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 1.06299 m³/h with 2 decimals of accuracy
09:51:46	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 1.771656
09:51:46	[V][pulse_meter:096]: New pulse, delta: 33866622 µs, count: 1, width: 33866624.00000 µs
09:51:46	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.14000 m³ with 2 decimals of accuracy
09:51:46	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 114.000000
09:51:57	[D][sensor:099]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Sending state -56.00000 dBm with 0 decimals of accuracy
09:51:57	[V][sensor:045]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Received new state -56.000000
09:52:11	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 2.09061 L/min with 1 decimals of accuracy
09:52:11	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 2.090607
09:52:11	[V][pulse_meter:096]: New pulse, delta: 28699804 µs, count: 1, width: 28699804.00000 µs
09:52:11	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 260.00000 L with 0 decimals of accuracy
09:52:11	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 260.000000
09:52:18	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 545454.56250 m³/h with 2 decimals of accuracy
09:52:18	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 909090.937500
09:52:18	[V][pulse_meter:096]: New pulse, delta: 66 µs, count: 1, width: 66.00000 µs
09:52:18	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.16000 m³ with 2 decimals of accuracy
09:52:18	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 116.000000
09:52:18	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 1.13616 m³/h with 2 decimals of accuracy
09:52:18	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 1.893602
09:52:18	[V][pulse_meter:096]: New pulse, delta: 31685650 µs, count: 1, width: 31685650.00000 µs
09:52:18	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.15000 m³ with 2 decimals of accuracy
09:52:18	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 115.000000
09:52:21	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 5.93507 L/min with 1 decimals of accuracy
09:52:21	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 5.935075
09:52:21	[V][pulse_meter:096]: New pulse, delta: 10109392 µs, count: 1, width: 10109392.00000 µs
09:52:21	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 261.00000 L with 0 decimals of accuracy
09:52:21	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 261.000000
09:52:34	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 4.53024 L/min with 1 decimals of accuracy
09:52:34	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 4.530243
09:52:34	[V][pulse_meter:096]: New pulse, delta: 13244322 µs, count: 1, width: 13244322.00000 µs
09:52:34	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 262.00000 L with 0 decimals of accuracy
09:52:34	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 262.000000
09:52:42	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 7.67169 L/min with 1 decimals of accuracy
09:52:42	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 7.671692
09:52:42	[V][pulse_meter:096]: New pulse, delta: 7820960 µs, count: 1, width: 7820960.00000 µs
09:52:42	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 263.00000 L with 0 decimals of accuracy
09:52:42	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 263.000000
09:52:48	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 1.19552 m³/h with 2 decimals of accuracy
09:52:48	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 1.992531
09:52:48	[V][pulse_meter:096]: New pulse, delta: 30112453 µs, count: 1, width: 30112452.00000 µs
09:52:48	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.17000 m³ with 2 decimals of accuracy
09:52:48	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 117.000000
09:52:57	[D][sensor:099]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Sending state -54.00000 dBm with 0 decimals of accuracy
09:52:57	[V][sensor:045]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Received new state -54.000000
09:53:07	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 2.43328 L/min with 1 decimals of accuracy
09:53:07	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 2.433275
09:53:07	[V][pulse_meter:096]: New pulse, delta: 24658125 µs, count: 1, width: 24658124.00000 µs
09:53:07	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 264.00000 L with 0 decimals of accuracy
09:53:07	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 264.000000
09:53:21	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 1.07444 m³/h with 2 decimals of accuracy
09:53:21	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 1.790735
09:53:21	[V][pulse_meter:096]: New pulse, delta: 33505795 µs, count: 1, width: 33505796.00000 µs
09:53:21	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.18000 m³ with 2 decimals of accuracy
09:53:21	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 118.000000
09:53:39	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 1.84848 L/min with 1 decimals of accuracy
09:53:39	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 1.848481
09:53:39	[V][pulse_meter:096]: New pulse, delta: 32459078 µs, count: 1, width: 32459078.00000 µs
09:53:39	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 265.00000 L with 0 decimals of accuracy
09:53:39	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 265.000000
09:53:57	[D][sensor:099]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Sending state -54.00000 dBm with 0 decimals of accuracy
09:53:57	[V][sensor:045]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Received new state -54.000000
09:54:07	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 2.16302 L/min with 1 decimals of accuracy
09:54:07	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 2.163021
09:54:07	[V][pulse_meter:096]: New pulse, delta: 27738978 µs, count: 1, width: 27738978.00000 µs
09:54:07	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 266.00000 L with 0 decimals of accuracy
09:54:07	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 266.000000
09:54:13	[D][sensor:099]: 'Utility Meter - Gas - Volume Flow Rate': Sending state 0.70253 m³/h with 2 decimals of accuracy
09:54:13	[V][sensor:045]: 'Utility Meter - Gas - Volume Flow Rate': Received new state 1.170887
09:54:13	[V][pulse_meter:096]: New pulse, delta: 51243188 µs, count: 1, width: 51243188.00000 µs
09:54:13	[D][sensor:099]: 'Utility Meter - Gas - Total Consumption': Sending state 1.19000 m³ with 2 decimals of accuracy
09:54:13	[V][sensor:045]: 'Utility Meter - Gas - Total Consumption': Received new state 119.000000
09:54:43	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 1.68797 L/min with 1 decimals of accuracy
09:54:43	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 1.687975
09:54:43	[V][pulse_meter:096]: New pulse, delta: 35545562 µs, count: 1, width: 35545560.00000 µs
09:54:43	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 267.00000 L with 0 decimals of accuracy
09:54:43	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 267.000000
09:54:56	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 4.44946 L/min with 1 decimals of accuracy
09:54:56	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 4.449455
09:54:56	[V][pulse_meter:096]: New pulse, delta: 13484797 µs, count: 1, width: 13484797.00000 µs
09:54:56	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 268.00000 L with 0 decimals of accuracy
09:54:56	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 268.000000
09:54:57	[D][sensor:099]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Sending state -54.00000 dBm with 0 decimals of accuracy
09:54:57	[V][sensor:045]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Received new state -54.000000
09:55:10	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 4.32429 L/min with 1 decimals of accuracy
09:55:10	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 4.324290
09:55:10	[V][pulse_meter:096]: New pulse, delta: 13875109 µs, count: 1, width: 13875109.00000 µs
09:55:10	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 269.00000 L with 0 decimals of accuracy
09:55:10	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 269.000000
09:55:50	[D][sensor:099]: 'Utility Meter - Water - Volume Flow Rate': Sending state 1.50732 L/min with 1 decimals of accuracy
09:55:50	[V][sensor:045]: 'Utility Meter - Water - Volume Flow Rate': Received new state 1.507320
09:55:50	[V][pulse_meter:096]: New pulse, delta: 39805741 µs, count: 1, width: 39805740.00000 µs
09:55:50	[D][sensor:099]: 'Utility Meter - Water - Total Consumption': Sending state 270.00000 L with 0 decimals of accuracy
09:55:50	[V][sensor:045]: 'Utility Meter - Water - Total Consumption': Received new state 270.000000
09:55:57	[D][sensor:099]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Sending state -55.00000 dBm with 0 decimals of accuracy
09:55:57	[V][sensor:045]: 'Utility Meter - Gas and Water - Wi-Fi Signal Strength': Received new state -55.000000
```

**Comment by boarder876** [comment]
is there anything new with the latest releases? I'm still on ESPHome 2023.7 where the pulse meter works, but I always need to handle two ESPHome versions which is annoying. 
Instead could i use...
```
external_components:
  - source: github://esphome/esphome@2023.7.1
    components: [pulse_meter]
```
...to use the old version of pulse meter together with latest releases of ESPHOME?

**Comment by TrentHouliston** [comment]
You could try https://github.com/esphome/esphome/pull/10619 I think that pull request actually fixes one of the major issues. Although I'm not sure if the author who was working on it was going to finish it. It's not merged yet either way

**Comment by boarder876** [comment]
> You could try [esphome/esphome#10619](https://github.com/esphome/esphome/pull/10619) I think that pull request actually fixes one of the major issues. Although I'm not sure if the author who was working on it was going to finish it. It's not merged yet either way

This solution is now working flawless for 4 weeks now without missing one single signal. The counter shows exact the same value as the meter itself.

**Comment by avbentem** [comment]
@boarder876, it seems that the PR you used [has been abandoned](https://github.com/esphome/esphome/pull/10619#issuecomment-3621729072). You may want to try https://github.com/esphome/esphome/pull/12360 instead. Which also takes a slightly different approach. (The latter works fine for me; I never tested the first.)

(Oh, my bad, you actually already commented there. Leaving this here for other future readers.)

**Comment by boarder876** [comment]
> [@boarder876](https://github.com/boarder876), it seems that the PR you used [has been abandoned](https://github.com/esphome/esphome/pull/10619#issuecomment-3621729072). You may want to try [esphome/esphome#12360](https://github.com/esphome/esphome/pull/12360) instead. Which also takes a slightly different approach. (The latter works fine for me; I never tested the first.)
> 
> (Oh, my bad, you actually already commented there. Leaving this here for other future readers.)

You're right, I've tested the lefevrer fix PR10616 for a long time without any issues. I also tested the PR12360 from @LucasCZE, which caused again some peak values ​​that should have been filtered out. I've flashed again the lefevrer version from LucasCZW, which works perfect for me. I compared the two versions, but I'm not a experienced programmer and don't understand the difference well enough to understand where the problem might lie.

**Comment by avbentem** [comment]
Just for future readers, [as for](https://github.com/esphome/issues/issues/4807#issuecomment-3546578526):

> Instead could i use...
> 
> ```
> external_components:
>   - source: github://esphome/esphome@2023.7.1
>     components: [pulse_meter]
> ```
> 
> ...to use the old version of pulse meter together with latest releases of ESPHOME?

I did not test myself, but [it seems the answer is: yes](https://github.com/smarthomeshop/watermeterkit/issues/6#issuecomment-3795519546).

**Comment by LucasCZE** [comment]
FYI: The issue should be fixed in 2026.2.0

[[pulse_meter] Fix early edge detection](https://github.com/esphome/esphome/pull/12360)
